packages/ses/src/make-hardener.js

```diff
@@ -27,26 +27,80 @@ import {
   TypeError,
   WeakMap,
   WeakSet,
+  apply,
   arrayForEach,
+  defineProperty,
   freeze,
+  getOwnPropertyDescriptor,
   getOwnPropertyDescriptors,
   getPrototypeOf,
+  isInteger,
   isObject,
   objectHasOwnProperty,
   ownKeys,
+  seal,
   setAdd,
   setForEach,
   setHas,
+  toStringTagSymbol,
+  typedArrayPrototype,
   weakmapGet,
   weakmapSet,
   weaksetAdd,
   weaksetHas,
 } from './commons.js';
+import { assert } from './error/assert.js';
 
 /**
  * @typedef {import('../index.js').Harden} Harden
  */
 
+// Obtain the string tag accessor of of TypedArray so we can indirectly use the
+// TypedArray brand check it employs.
+const typedArrayToStringTag = getOwnPropertyDescriptor(
+  typedArrayPrototype,
+  toStringTagSymbol,
+);
+assert(typedArrayToStringTag);
+const getTypedArrayToStringTag = typedArrayToStringTag.get;
+assert(getTypedArrayToStringTag);
+
+// Exported for tests.
+/** @param {unknown} object */
+export const isTypedArray = object => {
+  // The object must pass a brand check or toStringTag will return undefined.
+  const tag = apply(getTypedArrayToStringTag, object, []);
+  return tag !== undefined;
+};
+
+/**
+ * @template T
+ * @param {ArrayLike<T>} array
+ */
+const freezeTypedArray = array => {
+  const descs = getOwnPropertyDescriptors(array);
+
+  // Downgrade writable expandos to readonly, even if non-configurable.
+  arrayForEach(ownKeys(descs), (/** @type {string | symbol} */ name) => {
+    const desc = descs[/** @type {string} */ (name)];
+    // The numbered properties are writable and non-configurable,
+    // and cannot be made non-writable by defineProperty.
+    // This is a strange behavior intrinsic to TypedArrays, but no more harmful
+    // than the mutability of properties of a hardened Map or Set,
+    // so we carve out this exceptional behavior.
+    const number = +String(name);
+    if (!isInteger(number)) {
```

Comment:
Add a comment that the spec doesn't allow defining own props outside of the range `0..length` because typed arrays are ["Integer-Indexed Exotic Objects"](https://tc39.es/ecma262/#sec-integer-indexed-exotic-objects), so we're guaranteed that indexed props are the ones from the TypedArray instrinsic.

Reply:
I'm seeing bizarre behavior on v8:
```
> Object.defineProperty(ta, 88, { value: 99 })
Uncaught TypeError: Invalid typed array index
    at Function.defineProperty (<anonymous>)
> Object.defineProperty(ta, 88.1, { value: 99 })
Uncaught TypeError: Invalid typed array index
    at Function.defineProperty (<anonymous>)
> Object.defineProperty(ta, -88.1, { value: 99 })
Uncaught TypeError: Invalid typed array index
    at Function.defineProperty (<anonymous>)
> Object.defineProperty(ta, '-88.1', { value: 99 })
Uncaught TypeError: Invalid typed array index
    at Function.defineProperty (<anonymous>)
> Object.defineProperty(ta, '-88.10', { value: 99 })
Uint8Array(0) []
> ta['-88.10']
99
```
Is this a v8 bug? I don't think it makes the code above incorrect, but it does mean that the spec may not be an accurate picture of what engines do at runtime.

Reply:
That's indeed weird!